pkg/util/dynamichelper/discovery/cache_fallback_discovery_client.go

```diff
@@ -31,15 +32,30 @@ type cacheFallbackDiscoveryClient struct {
 
 // ServerResourcesForGroupVersion returns the supported resources for a group and version.
 func (d *cacheFallbackDiscoveryClient) ServerResourcesForGroupVersion(groupVersion string) (*metav1.APIResourceList, error) {
+	var cacheErr error
 	liveResources, err := d.DiscoveryInterface.ServerResourcesForGroupVersion(groupVersion)
 	if err != nil {
 		filename := filepath.Join(groupVersion, "serverresources.json")
 		cachedResources := &metav1.APIResourceList{}
-		cacheErr := d.getCached(filename, cachedResources)
+		cacheErr = d.getCached(filename, cachedResources)
 		if cacheErr == nil {
 			return cachedResources, nil
 		}
-		d.log.Info(cacheErr)
+		d.log.Infof("discovery cache failed. Fallback to live dynamic client. Error: %s", cacheErr)
+	}
+	// HACK:
+	// If live discovery cache failed with http.StatusServiceUnavailable,
+	// potentially one of the APIService is not responding and unhealthy.
+	// We check if we were able to resolve this resource from cache. If cache
+	// fails too - we drop resource on the floor and return nil.
+	// This builds on the assumption that code consuming this function is checking
+	// for NPE.
+	// discovery_client.go and cached_discovery.go is sole consumers of this
+	// code and we do check for != nil there.
+	// If user is requesting particular resource in question error will be returned:
+	// "The groupKind 'test.example.com' was not found."
+	if errors.IsServiceUnavailable(err) && cacheErr != nil {
```

Comment:
My initial plan was to do something similar, but after looking into how dynamichelper was implemented I had concerns about doing this... Trying to remember exact reasons. Will update once/if I remember.

Reply:
I don't know if it's possible, but I would much rather see us not ask for a GV that we don't care about in the first place.

Reply:
Hmm, the difficulty with that approach is it makes us hard to discover the group name for unqualified kinds - i.e. resolve SRE input of "clusterversion" to "clusterversion.config.openshift.io".